test/cri-containerd/clone_test.go

```diff
@@ -0,0 +1,404 @@
+// actually saved.
+// It is the callers responsibility to clean the stop and remove the cloned
+// containers and pods.
+func createTemplateContainer(ctx context.Context, t *testing.T, client runtime.RuntimeServiceClient, templateSandboxRequest *runtime.RunPodSandboxRequest, templateContainerRequest *runtime.CreateContainerRequest) (templatePodID, templateContainerID string) {
+	templatePodID, templateContainerID = createPodAndContainer(ctx, t, client, templateSandboxRequest, templateContainerRequest)
+
+	// Send a context with deadline for waitForTemplateSave function
+	d := time.Now().Add(10 * time.Second)
+	ctx, cancel := context.WithDeadline(ctx, d)
+	defer cancel()
+	waitForTemplateSave(ctx, t, templatePodID)
+	return
+}
+
+// Creates a clone from the given template pod and container.
+// It is the callers responsibility to clean the stop and remove the cloned
+// containers and pods.
+func createClonedContainer(ctx context.Context, t *testing.T, client runtime.RuntimeServiceClient, templatePodID, templateContainerID string, cloneNumber int) (clonedPodID, clonedContainerID string) {
+	cloneSandboxRequest := getClonedPodConfig(cloneNumber, templatePodID)
+	cloneContainerRequest := getClonedContainerConfig(cloneNumber, templateContainerID)
+	clonedPodID, clonedContainerID = createPodAndContainer(ctx, t, client, cloneSandboxRequest, cloneContainerRequest)
+	return
+}
+
+// Runs a command inside given container and verifies if the command executes successfully.
+func verifyContainerExec(ctx context.Context, t *testing.T, client runtime.RuntimeServiceClient, containerID string) {
+	execCommand := []string{
+		"ping",
+		"www.microsoft.com",
+	}
+
+	execRequest := &runtime.ExecSyncRequest{
+		ContainerId: containerID,
+		Cmd:         execCommand,
+		Timeout:     20,
+	}
+
+	r := execSync(t, client, ctx, execRequest)
+	output := strings.TrimSpace(string(r.Stdout))
+	errorMsg := string(r.Stderr)
+	exitCode := int(r.ExitCode)
+
+	if exitCode != 0 || len(errorMsg) != 0 {
+		t.Fatalf("Failed execution inside container %s with error: %s, exitCode: %d", containerID, errorMsg, exitCode)
+	} else {
+		t.Logf("Exec(container: %s) stdout: %s, stderr: %s, exitCode: %d\n", containerID, output, errorMsg, exitCode)
+	}
+}
+
+// A simple test to just create a template container and then create one
+// cloned container from that template.
+func Test_CloneContainer_WCOW(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	client := newTestRuntimeClient(t)
+
+	pullRequiredImages(t, []string{imageWindowsNanoserver})
+
+	templatePodID, templateContainerID := createTemplateContainer(ctx, t, client, getTemplatePodConfig("templatepod"), getTemplateContainerConfig("templatecontainer"))
+	defer removePodSandbox(t, client, ctx, templatePodID)
+	defer stopPodSandbox(t, client, ctx, templatePodID)
+	defer removeContainer(t, client, ctx, templateContainerID)
+	defer stopContainer(t, client, ctx, templateContainerID)
```

Comment:
I don't think this will be sufficient. In the event of a panic in `createTemplateContainer`, these defers would not be run, in which case we could leak a pod and/or container. The defer statements need to be immediately after each create/start of a pod/container to ensure that in the event of say "start pod" panicking, the previously set pod removal defer would still be run and the pod would be properly cleaned up.

Reply:
Hmm. This is a good point. The use of `t.Fatalf` in `runPodSandbox` and `createContainer` functions make it difficult to keep the cleanup part in the main test function. I have changed my functions to deal with this but I think we should replace the `t.Fatalf` with some error so that the caller can deal with the cleanup.